src/v/storage/segment_utils.cc

```diff
@@ -470,42 +481,42 @@ ss::future<> self_compact_segment(
-                vlog(stlog.info, "detected {} is already compacted", idx_path);
-                return ss::now();
-            }
-            case compacted_index::recovery_state::nonrecovered:
-                return do_self_compact_segment(s, cfg, pb);
-            case compacted_index::recovery_state::missing:
-                [[fallthrough]];
-            case compacted_index::recovery_state::needsrebuild: {
-                vlog(
-                  stlog.warn,
-                  "Detected corrupt or missing index file:{}, recovering...",
-                  idx_path);
-                pb.corrupted_compaction_index();
-                return s->read_lock()
-                  .then([s, cfg, &pb, idx_path](ss::rwlock::holder h) {
-                      return rebuild_compaction_index(
-                        create_segment_full_reader(s, cfg, pb, std::move(h)),
-                        idx_path,
-                        cfg);
-                  })
-                  .then([s, cfg, &pb, idx_path] {
-                      vlog(
-                        stlog.info,
-                        "recovered index: {}, attempting compaction again",
-                        idx_path);
-                      return self_compact_segment(s, cfg, pb);
-                  });
-            }
-            default:
-                __builtin_unreachable();
-            }
-        })
+      .then([idx_path, s, cfg, &pb, &readers_cache](
+              compacted_index::recovery_state state) mutable {
+          switch (state) {
+          case compacted_index::recovery_state::recovered: {
+              vlog(stlog.info, "detected {} is already compacted", idx_path);
+              return ss::now();
+          }
+          case compacted_index::recovery_state::nonrecovered:
+              return do_self_compact_segment(s, cfg, pb, readers_cache);
+          case compacted_index::recovery_state::missing:
+              [[fallthrough]];
+          case compacted_index::recovery_state::needsrebuild: {
+              vlog(
+                stlog.warn,
+                "Detected corrupt or missing index file:{}, recovering...",
+                idx_path);
+              pb.corrupted_compaction_index();
+              return s->read_lock()
+                .then([s, cfg, &pb, idx_path](ss::rwlock::holder h) {
+                    return rebuild_compaction_index(
+                      create_segment_full_reader(s, cfg, pb, std::move(h)),
+                      idx_path,
+                      cfg);
+                })
+                .then([s, cfg, &pb, idx_path, &readers_cache] {
+                    vlog(
+                      stlog.info,
+                      "recovered index: {}, attempting compaction again",
+                      idx_path);
+                    return self_compact_segment(s, cfg, pb, readers_cache);
+                });
+          }
+          default:
+              __builtin_unreachable();
```

Comment:
nit: builtin unreachable should not be in the default case. it prevents compiler from catching missing enums.

Reply:
this change comes from reformatting, I will move this into different commit